In [1]:
import os
import glob
import pandas as pd
import numpy as np

pd.options.display.max_columns = 40

In [2]:
# Downloads data from Kaggle and extracts it from the zip files

# !kaggle competitions download -c recruit-restaurant-visitor-forecasting -p .\Data
# !kaggle datasets download huntermcgushion/rrv-weather-data -p Data\Weather -- unzip

# from zipfile import ZipFile

# for filename in os.listdir('Data'):
#     if filename.endswith('.zip'):
#         with ZipFile('Data\\' + filename) as zip_obj:
#             zip_obj.extractall('Data')

In [3]:
data = {}
for file in os.listdir('Data'):
    if file.endswith('.csv'):
        data[file.split('.')[0]] = pd.read_csv('Data/' + file)
data.keys()

dict_keys(['air_reserve', 'air_store_info', 'air_visit_data', 'date_info', 'hpg_reserve', 'hpg_store_info', 'sample_submission', 'store_id_relation'])

In [4]:
data['air_reserve'].head()

air_store_id       visit_datetime     reserve_datetime  \
0  air_877f79706adbfb06  2016-01-01 19:00:00  2016-01-01 16:00:00   
1  air_db4b38ebe7a7ceff  2016-01-01 19:00:00  2016-01-01 19:00:00   
2  air_db4b38ebe7a7ceff  2016-01-01 19:00:00  2016-01-01 19:00:00   
3  air_877f79706adbfb06  2016-01-01 20:00:00  2016-01-01 16:00:00   
4  air_db80363d35f10926  2016-01-01 20:00:00  2016-01-01 01:00:00   

   reserve_visitors  
0                 1  
1                 3  
2                 6  
3                 2  
4                 5

In [5]:
data['air_store_info'].head()

air_store_id  air_genre_name                 air_area_name  \
0  air_0f0cdeee6c9bf3d7  Italian/French  Hyōgo-ken Kōbe-shi Kumoidōri   
1  air_7cc17a324ae5c7dc  Italian/French  Hyōgo-ken Kōbe-shi Kumoidōri   
2  air_fee8dcf4d619598e  Italian/French  Hyōgo-ken Kōbe-shi Kumoidōri   
3  air_a17f0778617c76e2  Italian/French  Hyōgo-ken Kōbe-shi Kumoidōri   
4  air_83db5aff8f50478e  Italian/French  Tōkyō-to Minato-ku Shibakōen   

    latitude   longitude  
0  34.695124  135.197852  
1  34.695124  135.197852  
2  34.695124  135.197852  
3  34.695124  135.197852  
4  35.658068  139.751599

In [6]:
data['air_store_info']['city'] = [x[0] for x in data['air_store_info']['air_area_name'].str.split()]

In [7]:
data['air_visit_data'].head()

air_store_id  visit_date  visitors
0  air_ba937bf13d40fb24  2016-01-13        25
1  air_ba937bf13d40fb24  2016-01-14        32
2  air_ba937bf13d40fb24  2016-01-15        29
3  air_ba937bf13d40fb24  2016-01-16        22
4  air_ba937bf13d40fb24  2016-01-18         6

In [8]:
data['air_visit_data'].columns = ['air_store_id', 'calendar_date', 'visitors']

In [9]:
data['date_info'].head()

calendar_date day_of_week  holiday_flg
0    2016-01-01      Friday            1
1    2016-01-02    Saturday            1
2    2016-01-03      Sunday            1
3    2016-01-04      Monday            0
4    2016-01-05     Tuesday            0

In [10]:
data['hpg_reserve'].head()

hpg_store_id       visit_datetime     reserve_datetime  \
0  hpg_c63f6f42e088e50f  2016-01-01 11:00:00  2016-01-01 09:00:00   
1  hpg_dac72789163a3f47  2016-01-01 13:00:00  2016-01-01 06:00:00   
2  hpg_c8e24dcf51ca1eb5  2016-01-01 16:00:00  2016-01-01 14:00:00   
3  hpg_24bb207e5fd49d4a  2016-01-01 17:00:00  2016-01-01 11:00:00   
4  hpg_25291c542ebb3bc2  2016-01-01 17:00:00  2016-01-01 03:00:00   

   reserve_visitors  
0                 1  
1                 3  
2                 2  
3                 5  
4                13

In [11]:
data['hpg_store_info'].head()

hpg_store_id  hpg_genre_name                  hpg_area_name  \
0  hpg_6622b62385aec8bf  Japanese style  Tōkyō-to Setagaya-ku Taishidō   
1  hpg_e9e068dd49c5fa00  Japanese style  Tōkyō-to Setagaya-ku Taishidō   
2  hpg_2976f7acb4b3a3bc  Japanese style  Tōkyō-to Setagaya-ku Taishidō   
3  hpg_e51a522e098f024c  Japanese style  Tōkyō-to Setagaya-ku Taishidō   
4  hpg_e3d0e1519894f275  Japanese style  Tōkyō-to Setagaya-ku Taishidō   

    latitude   longitude  
0  35.643675  139.668221  
1  35.643675  139.668221  
2  35.643675  139.668221  
3  35.643675  139.668221  
4  35.643675  139.668221

In [12]:
data['store_id_relation'].head()

air_store_id          hpg_store_id
0  air_63b13c56b7201bd9  hpg_4bc649e72e2a239a
1  air_a24bf50c3e90d583  hpg_c34b496d0305a809
2  air_c7f78b4f3cba33ff  hpg_cd8ae0d9bbd58ff9
3  air_947eb2cae4f3e8f2  hpg_de24ea49dc25d6b8
4  air_965b2e0cf4119003  hpg_653238a84804d8e7

In [13]:
print(data['air_store_info'].shape)
print(data['hpg_store_info'].shape)
print(data['store_id_relation'].shape)

(829, 6)
(4690, 5)
(150, 2)


Only 150 of the restaurants are in both datasets. I am trying to predict the number of visitors to each restaurant in the Air dataset so I might use the Air and HPG reserve data to create a column of number of reservations per day, but they do not need to be merged to the visit data.

The people who scraped and collected the weather data also did the work of calculating the closest weather station for each restaurant and appended it to the store data.

In [14]:
air_store_weather = pd.read_csv('Data/Weather/air_store_info_with_nearest_active_station.csv')

In [15]:
features = pd.read_csv('Data/Weather/feature_manifest.csv')

In [16]:
features.describe()

avg_temperature  high_temperature  low_temperature  precipitation  \
count      1663.000000       1663.000000      1663.000000    1663.000000   
mean          0.558560          0.558508         0.558508       0.744709   
std           0.495644          0.495598         0.495598       0.407874   
min           0.000000          0.000000         0.000000       0.000000   
25%           0.000000          0.000000         0.000000       0.578337   
50%           0.998066          0.996132         0.996132       1.000000   
75%           1.000000          1.000000         1.000000       1.000000   
max           1.000000          1.000000         1.000000       1.000000   

       hours_sunlight  solar_radiation  deepest_snowfall  total_snowfall  \
count     1663.000000      1663.000000       1663.000000     1663.000000   
mean         0.504173         0.028848          0.144023        0.139755   
std          0.498516         0.167386          0.341394        0.339987   
min          0.000000         0.000000          0.000000        0.000000   
25%          0.000000         0.000000          0.000000        0.000000   
50%          0.965184         0.000000          0.000000        0.000000   
75%          1.000000         0.000000          0.000000        0.000000   
max          1.000000         1.000000          1.000000        1.000000   

       avg_wind_speed  avg_vapor_pressure  avg_local_pressure  avg_humidity  \
count     1663.000000         1663.000000         1663.000000   1663.000000   
mean         0.550155            0.093494            0.094267      0.093495   
std          0.490407            0.290432            0.292093      0.290433   
min          0.000000            0.000000            0.000000      0.000000   
25%          0.000000            0.000000            0.000000      0.000000   
50%          0.965184            0.000000            0.000000      0.000000   
75%          0.998066            0.000000            0.000000      0.000000   
max          1.000000            1.000000            1.000000      1.000000   

       avg_sea_pressure  cloud_cover  
count       1663.000000  1663.000000  
mean           0.090785     0.035372  
std            0.287363     0.184538  
min            0.000000     0.000000  
25%            0.000000     0.000000  
50%            0.000000     0.000000  
75%            0.000000     0.000000  
max            1.000000     1.000000

In [17]:
features[(features.avg_temperature==1) & (features.precipitation==1)].describe()

avg_temperature  high_temperature  low_temperature  precipitation  \
count            568.0        568.000000       568.000000          568.0   
mean               1.0          0.999949         0.999949            1.0   
std                0.0          0.000478         0.000478            0.0   
min                1.0          0.992263         0.992263            1.0   
25%                1.0          1.000000         1.000000            1.0   
50%                1.0          1.000000         1.000000            1.0   
75%                1.0          1.000000         1.000000            1.0   
max                1.0          1.000000         1.000000            1.0   

       hours_sunlight  solar_radiation  deepest_snowfall  total_snowfall  \
count      568.000000            568.0        568.000000      568.000000   
mean         0.949519              0.0          0.256719        0.256664   
std          0.216440              0.0          0.436843        0.436751   
min          0.000000              0.0          0.000000        0.000000   
25%          0.998066              0.0          0.000000        0.000000   
50%          1.000000              0.0          0.000000        0.000000   
75%          1.000000              0.0          0.991296        0.990329   
max          1.000000              0.0          1.000000        1.000000   

       avg_wind_speed  avg_vapor_pressure  avg_local_pressure  avg_humidity  \
count      568.000000               568.0               568.0         568.0   
mean         0.986668                 0.0                 0.0           0.0   
std          0.021250                 0.0                 0.0           0.0   
min          0.856867                 0.0                 0.0           0.0   
25%          0.982592                 0.0                 0.0           0.0   
50%          0.996132                 0.0                 0.0           0.0   
75%          1.000000                 0.0                 0.0           0.0   
max          1.000000                 0.0                 0.0           0.0   

       avg_sea_pressure  cloud_cover  
count             568.0        568.0  
mean                0.0          0.0  
std                 0.0          0.0  
min                 0.0          0.0  
25%                 0.0          0.0  
50%                 0.0          0.0  
75%                 0.0          0.0  
max                 0.0          0.0

Looking through the feature manifest which contains a 1 if that measurement is available for all days at a given weather station and a 0 if it not available for any days, I found that no stations measure everything all the time. To reduce the number of missing values, I will average the closest stations.

In [18]:
distances = (pd.read_csv('Data/Weather/air_station_distances.csv', index_col=0)
             .drop(['station_latitude', 'station_longitude'], axis=1))
retired_stations = distances[distances.index.str.contains('___')].index
distances.drop(retired_stations, inplace=True)
distances.head(2)

(34.6951242, 135.1978525)  \
station_id                                            
aichi__ai-xi-kana__isaai                 148.760474   
aichi__aki-kana__azo                     206.934423   

                          (35.6580681, 139.7515992)  \
station_id                                            
aichi__ai-xi-kana__isaai                 281.515494   
aichi__aki-kana__azo                     222.009692   

                          (33.5892157, 130.3928134)  \
station_id                                            
aichi__ai-xi-kana__isaai                 607.092649   
aichi__aki-kana__azo                     666.658179   

                          (35.6465721, 139.6532473)  \
station_id                                            
aichi__ai-xi-kana__isaai                 272.519535   
aichi__aki-kana__azo                     213.104447   

                          (34.7053617, 135.5100252)  \
station_id                                            
aichi__ai-xi-kana__isaai                 122.458285   
aichi__aki-kana__azo                     178.827179   

                          (35.6617773, 139.7040506)  \
station_id                                            
aichi__ai-xi-kana__isaai                 277.332264   
aichi__aki-kana__azo                     217.995118   

                          (35.6995657, 139.6364379)  \
station_id                                            
aichi__ai-xi-kana__isaai                 272.027251   
aichi__aki-kana__azo                     213.408094   

                          (34.3862445, 132.4550176)  \
station_id                                            
aichi__ai-xi-kana__isaai                 399.040385   
aichi__aki-kana__azo                     460.497590   

                          (35.6721142, 139.7708253)  \
station_id                                            
aichi__ai-xi-kana__isaai                 283.484213   
aichi__aki-kana__azo                     224.117668   

                          (35.6938401, 139.7035494)  \
station_id                                            
aichi__ai-xi-kana__isaai                 277.889737   
aichi__aki-kana__azo                     218.995182   

                          (35.6414629, 139.6981712)  (34.7594526, 135.51686)  \
station_id                                                                     
aichi__ai-xi-kana__isaai                 276.447682               119.199686   
aichi__aki-kana__azo                     216.849903               176.867128   

                          (35.7140142, 139.4078431)  \
station_id                                            
aichi__ai-xi-kana__isaai                 252.042605   
aichi__aki-kana__azo                     194.430301   

                          (33.5913576, 130.4148783)  \
station_id                                            
aichi__ai-xi-kana__isaai                 605.079289   
aichi__aki-kana__azo                     664.625726   

                          (34.7108955, 137.7259397)  \
station_id                                            
aichi__ai-xi-kana__isaai                 109.342984   
aichi__aki-kana__azo                      49.780331   

                          (43.0554601, 141.3409556)  (35.6749184, 139.736775)  \
station_id                                                                      
aichi__ai-xi-kana__isaai                 957.952227                280.494658   
aichi__aki-kana__azo                     948.288407                221.251709   

                          (35.6706505, 139.7718614)  \
station_id                                            
aichi__ai-xi-kana__isaai                 283.550369   
aichi__aki-kana__azo                     224.161850   

                          (34.7569503, 134.8411775)  \
station_id                                            
aichi__ai-xi-kana__isaai                 177.064122   
aichi__aki-kana__azo                     237.724097   

                          (35.6940027, 139.7535951)  ...  \
stati

In [19]:
def closest_five_stations(srs):
    smallest = srs.sort_values()[0:5].index
    is_lowest = pd.Series(0, index=srs.index)
    is_lowest.loc[smallest] = 1
    return is_lowest

closest = distances.apply(lambda srs: closest_five_stations(srs))

In [20]:
path = 'Data/Weather/1-1-16_5-31-17_Weather/'
all_files = glob.glob(os.path.join(path, '*.csv'))
weather = pd.concat((pd.read_csv(f).assign(station_id=f.split('\\')[-1].split('.')[0]) for f in all_files),
                    ignore_index=True)
weather['calendar_date'] = pd.to_datetime(weather['calendar_date'])
weather = weather.set_index(['calendar_date', 'station_id']).sort_index()

In [21]:
weather.head()

avg_temperature  \
calendar_date station_id                                                
2016-01-01    aichi__ai-xi-kana__isaai                            6.0   
              aichi__aki-kana__azo                                NaN   
              aichi__centrair-kana__centrair                      8.9   
              aichi__chapel-mountain-kana__chaw-yama              NaN   
              aichi__gamagori-kana__gamagori                      6.6   

                                                      high_temperature  \
calendar_date station_id                                                 
2016-01-01    aichi__ai-xi-kana__isaai                            11.0   
              aichi__aki-kana__azo                                 NaN   
              aichi__centrair-kana__centrair                      11.1   
              aichi__chapel-mountain-kana__chaw-yama               NaN   
              aichi__gamagori-kana__gamagori                      11.2   

                                                      low_temperature  \
calendar_date station_id                                                
2016-01-01    aichi__ai-xi-kana__isaai                            0.7   
              aichi__aki-kana__azo                                NaN   
              aichi__centrair-kana__centrair                      5.9   
              aichi__chapel-mountain-kana__chaw-yama              NaN   
              aichi__gamagori-kana__gamagori                      2.5   

                                                      precipitation  \
calendar_date station_id                                              
2016-01-01    aichi__ai-xi-kana__isaai                          0.0   
              aichi__aki-kana__azo                              0.0   
              aichi__centrair-kana__centrair                    0.0   
              aichi__chapel-mountain-kana__chaw-yama            0.0   
              aichi__gamagori-kana__gamagori                    0.0   

                                                      hours_sunlight  \
calendar_date station_id                                               
2016-01-01    aichi__ai-xi-kana__isaai                           9.1   
              aichi__aki-kana__azo                               NaN   
              aichi__centrair-kana__centrair                     NaN   
              aichi__chapel-mountain-kana__chaw-yama             NaN   
              aichi__gamagori-kana__gamagori                     9.4   

                                                      solar_radiation  \
calendar_date station_id                                                
2016-01-01    aichi__ai-xi-kana__isaai                            NaN   
              aichi__aki-kana__azo                                NaN   
              aichi__centrair-kana__centrair                      NaN   
              aichi__chapel-mountain-kana__chaw-yama              NaN   
              aichi__gamagori-kana__gamagori                      NaN   

                                                      deepest_snowfall  \
calendar_date station_id                                                 
2016-01-01    aichi__ai-xi-kana__isaai                             NaN   
              aichi__aki-kana__azo                                 NaN   
              aichi__centrair-kana__centrair                       NaN   
              aichi__chapel-mountain-kana__chaw-yama               NaN   
              aichi__gamagori-kana__gamagori                       NaN   

                                                      total_snowfall  \
calendar_date station_id                                               
2016-01-01    aichi__ai-xi-kana__isaai                           NaN   
              aichi__aki-kana__azo                               NaN   
              aichi__centrair-kana__centrair                     NaN   
              aichi__chapel-mountain-kana__chaw-yama             NaN   
              aichi__gamagori-kana__gamagori 

In [22]:
dates = weather.index.get_level_values('calendar_date').unique()
locations = closest.columns
measurements = weather.columns
new_index = pd.MultiIndex.from_product([locations, dates], names=['location', 'calendar_date'])
average_weather = pd.DataFrame(0, index=new_index, columns=measurements)
average_weather.sort_index(level=[0,1], inplace=True)

In [23]:
for date in weather.index.get_level_values('calendar_date').unique():
    current_dates_weather = weather.loc[date]
    for location in average_weather.index.get_level_values('location').unique():
        locations_to_average = closest[closest.loc[:,location].astype('bool')].index
        average_weather.loc[(location, date), :] = (current_dates_weather
                                                    .loc[locations_to_average, :]
                                                    .mean())

In [24]:
average_weather.head()

avg_temperature  high_temperature  \
location                  calendar_date                                      
(33.2119674, 130.5579082) 2016-01-01            4.466667         11.066667   
                          2016-01-02            9.066667         14.466667   
                          2016-01-03           11.000000         16.266667   
                          2016-01-04            8.866667         15.200000   
                          2016-01-05            8.266667         11.633333   

                                         low_temperature  precipitation  \
location                  calendar_date                                   
(33.2119674, 130.5579082) 2016-01-01           -0.400000            0.0   
                          2016-01-02            4.266667            0.2   
                          2016-01-03            6.333333            0.0   
                          2016-01-04            4.533333            0.0   
                          2016-01-05            6.600000            3.6   

                                         hours_sunlight  solar_radiation  \
location                  calendar_date                                    
(33.2119674, 130.5579082) 2016-01-01           8.366667              NaN   
                          2016-01-02           0.900000              NaN   
                          2016-01-03           0.366667              NaN   
                          2016-01-04           5.733333              NaN   
                          2016-01-05           0.000000              NaN   

                                         deepest_snowfall  total_snowfall  \
location                  calendar_date                                     
(33.2119674, 130.5579082) 2016-01-01                  NaN             NaN   
                          2016-01-02                  NaN             NaN   
                          2016-01-03                  NaN             NaN   
                          2016-01-04                  NaN             NaN   
                          2016-01-05                  NaN             NaN   

                                         avg_wind_speed  avg_vapor_pressure  \
location                  calendar_date                                       
(33.2119674, 130.5579082) 2016-01-01           0.800000                 NaN   
                          2016-01-02           0.700000                 NaN   
                          2016-01-03           0.700000                 NaN   
                          2016-01-04           0.666667                 NaN   
                          2016-01-05           1.300000                 NaN   

                                         avg_local_pressure  avg_humidity  \
location                  calendar_date                                     
(33.2119674, 130.5579082) 2016-01-01                    NaN           NaN   
                          2016-01-02                    NaN           NaN   
                          2016-01-03                    NaN           NaN   
                          2016-01-04                    NaN           NaN   
                          2016-01-05                    NaN           NaN   

                                         avg_sea_pressure  cloud_cover  
location                  calendar_date                                 
(33.2119674, 130.5579082) 2016-01-01                  NaN          NaN  
                          2016-01-02                  NaN          NaN  
                          2016-01-03                  NaN          NaN  
                          2016-01-04                  NaN          NaN  
                          2016-01-05                  NaN          NaN

In [25]:
air_store_weather.set_index('air_store_id', inplace=True)
air_store_weather['location'] = ('(' + air_store_weather['latitude_str'].str.strip('\"') + ', ' +
                                 air_store_weather['longitude_str'].str.strip('\"') + ')')

Before merging all the store info and weather data, I need to bring in the test data so that the merge and cleaning must only be done once.

In [26]:
test = data['sample_submission'].copy()
test['air_store_id'] = test['id'].str.slice(0, 20)
test['calendar_date'] = test['id'].str.slice(21)
test['visitors'] = np.nan
test['is_test'] = True

In [27]:
test.head()

id  visitors          air_store_id  \
0  air_00a91d42b08b08d9_2017-04-23       NaN  air_00a91d42b08b08d9   
1  air_00a91d42b08b08d9_2017-04-24       NaN  air_00a91d42b08b08d9   
2  air_00a91d42b08b08d9_2017-04-25       NaN  air_00a91d42b08b08d9   
3  air_00a91d42b08b08d9_2017-04-26       NaN  air_00a91d42b08b08d9   
4  air_00a91d42b08b08d9_2017-04-27       NaN  air_00a91d42b08b08d9   

  calendar_date  is_test  
0    2017-04-23     True  
1    2017-04-24     True  
2    2017-04-25     True  
3    2017-04-26     True  
4    2017-04-27     True

In [28]:
air_visit_data = pd.concat((data['air_visit_data'], test.drop('id', axis='columns')), sort=True)
air_visit_data['is_test'].fillna(False, inplace=True)

In [29]:
air_visit_store = (air_visit_data
                   .merge(data['air_store_info'], how='left', on='air_store_id')
                   .merge(air_store_weather['location'], how='left', on='air_store_id')
                   .merge(data['date_info'], how='left', on='calendar_date'))

In [30]:
air_visit_store['calendar_date'] = pd.to_datetime(air_visit_store['calendar_date'])

In [31]:
air_visit_weather = (air_visit_store
                     .merge(average_weather, how='left', on=['location', 'calendar_date'])
                     .set_index(['air_store_id', 'calendar_date']))

In [32]:
air_visit_weather.head()

is_test  visitors air_genre_name  \
air_store_id         calendar_date                                     
air_ba937bf13d40fb24 2016-01-13       False      25.0     Dining bar   
                     2016-01-14       False      32.0     Dining bar   
                     2016-01-15       False      29.0     Dining bar   
                     2016-01-16       False      22.0     Dining bar   
                     2016-01-18       False       6.0     Dining bar   

                                                   air_area_name   latitude  \
air_store_id         calendar_date                                            
air_ba937bf13d40fb24 2016-01-13     Tōkyō-to Minato-ku Shibakōen  35.658068   
                     2016-01-14     Tōkyō-to Minato-ku Shibakōen  35.658068   
                     2016-01-15     Tōkyō-to Minato-ku Shibakōen  35.658068   
                     2016-01-16     Tōkyō-to Minato-ku Shibakōen  35.658068   
                     2016-01-18     Tōkyō-to Minato-ku Shibakōen  35.658068   

                                     longitude      city  \
air_store_id         calendar_date                         
air_ba937bf13d40fb24 2016-01-13     139.751599  Tōkyō-to   
                     2016-01-14     139.751599  Tōkyō-to   
                     2016-01-15     139.751599  Tōkyō-to   
                     2016-01-16     139.751599  Tōkyō-to   
                     2016-01-18     139.751599  Tōkyō-to   

                                                     location day_of_week  \
air_store_id         calendar_date                                          
air_ba937bf13d40fb24 2016-01-13     (35.6580681, 139.7515992)   Wednesday   
                     2016-01-14     (35.6580681, 139.7515992)    Thursday   
                     2016-01-15     (35.6580681, 139.7515992)      Friday   
                     2016-01-16     (35.6580681, 139.7515992)    Saturday   
                     2016-01-18     (35.6580681, 139.7515992)      Monday   

                                    holiday_flg  avg_temperature  \
air_store_id         calendar_date                                 
air_ba937bf13d40fb24 2016-01-13               0         4.833333   
                     2016-01-14               0         6.566667   
                     2016-01-15               0         6.033333   
                     2016-01-16               0         6.700000   
                     2016-01-18               0         3.833333   

                                    high_temperature  low_temperature  \
air_store_id         calendar_date                                      
air_ba937bf13d40fb24 2016-01-13             9.866667        -0.700000   
                     2016-01-14            11.233333         0.366667   
                     2016-01-15            10.566667         2.800000   
                     2016-01-16            11.400000         2.733333   
                     2016-01-18             8.433333         0.400000   

                                    precipitation  hours_sunlight  \
air_store_id         calendar_date                                  
air_ba937bf13d40fb24 2016-01-13               0.0            8.75   
                     2016-01-14               0.0            9.30   
                     2016-01-15               0.0            8.05   
                     2016-01-16               0.0            9.25   
                     2016-01-18              53.4            0.90   

                                    solar_radiation  deepest_snowfall  \
air_store_id         calendar_date                                      
air_ba937bf13d40fb24 2016-01-13               10.86               NaN   
                     2016-01-14               12.09               NaN   
                     2016-01-15               11.67               NaN   
                     2016-01-16               12.41               NaN   
                     2016-01-18                2.40               6.0   

                               

In [33]:
drop_cols = ['latitude', 'longitude', 'location']
df = air_visit_weather.drop(columns=drop_cols)
df.reset_index(inplace=True)

In [34]:
df.isna().sum().sort_values(ascending=False)

total_snowfall        264140
deepest_snowfall      262644
solar_radiation        33786
visitors               32019
cloud_cover            22993
avg_humidity            9636
avg_vapor_pressure      9636
avg_local_pressure      9247
avg_sea_pressure        9247
avg_wind_speed             6
city                       0
calendar_date              0
is_test                    0
air_genre_name             0
air_area_name              0
low_temperature            0
day_of_week                0
holiday_flg                0
avg_temperature            0
high_temperature           0
precipitation              0
hours_sunlight             0
air_store_id               0
dtype: int64

The two snowfall features have more than 90% missing so I will drop them. The rest I will fill based on the city and the date.

In [35]:
df.drop(columns=['total_snowfall', 'deepest_snowfall'], inplace=True)

In [36]:
def fill_missing_weather(row):
    for col in missing_cols:
        if np.isnan(row[col]):
            row[col] = fill_values_city_date[col][(row['city'], row['calendar_date'])]
            if np.isnan(row[col]):
                row[col] = fill_values_date[col][row['calendar_date']]
    return row

missing_cols = ['solar_radiation', 'visitors','cloud_cover', 'avg_humidity',
                'avg_vapor_pressure', 'avg_local_pressure', 'avg_sea_pressure', 'avg_wind_speed']
fill_values_city_date = df.groupby(by=['city', 'calendar_date'])[missing_cols].mean().to_dict()
fill_values_date = df.groupby(by=['calendar_date'])[missing_cols].mean().to_dict()
df.loc[df.isna().any(axis=1)] = df[df.isna().any(axis=1)].apply(fill_missing_weather, axis=1)

In [37]:
df.isna().sum().sort_values(ascending=False)

visitors              32019
cloud_cover               0
avg_temperature           0
calendar_date             0
is_test                   0
air_genre_name            0
air_area_name             0
city                      0
day_of_week               0
holiday_flg               0
high_temperature          0
avg_sea_pressure          0
low_temperature           0
precipitation             0
hours_sunlight            0
solar_radiation           0
avg_wind_speed            0
avg_vapor_pressure        0
avg_local_pressure        0
avg_humidity              0
air_store_id              0
dtype: int64

In [38]:
df.to_csv('Data/Processed/air_data.csv')
average_weather.to_csv('Data/Processed/average_weather.csv')